llumnix/backends/vllm/migrate_backend.py

```diff
@@ -0,0 +1,257 @@
+
+class MigrateBackendBase(ABC):
+    @abstractmethod
+    def init_col(self, name, world_size, rank) -> None:
+        raise NotImplementedError
+
+    @abstractmethod
+    def destory_col(self) -> None:
+        raise NotImplementedError
+
+    @abstractmethod
+    def warmup(self) -> None:
+        raise NotImplementedError
+
+    @abstractmethod
+    def migrate_cache(self, src_handle, src_blocks: List[int], dst_blocks: List[int]) -> None:
+        raise NotImplementedError
+
+    @abstractmethod
+    def do_send(self, dst_handle, blocks: List[int]):
+        raise NotImplementedError
+
+    @abstractmethod
+    def do_recv(self, src_handle, blocks: List[int]):
+        raise NotImplementedError
+
+@ray.remote(num_cpus=0)
+class ProxyActor:
+    def exec_method(self, is_driver_worker, handle, *args, **kwargs):
+        try:
+            if is_driver_worker:
+                ret = ray.get(handle.execute_engine_method.remote("execute_worker_method", *args, **kwargs))
+            else:
+                ret = ray.get(handle.execute_method.remote(*args, **kwargs))
+        # pylint: disable=try-except-raise
+        except:
+            raise
+
+        return ret
+
+class RPCMigrateBackend(MigrateBackendBase):
+    def __init__(self, migrate_config: MigrationConfig, cache_engine: CacheEngine,  worker_rank, worker_handle_list, \
+                  scheduling_strategy, dtype, is_driver_worker, gpu_cache) -> None:
+        super().__init__()
+
+        self.migrate_config = migrate_config
+        self.cache_engine = cache_engine
+
+        self.worker_rank = worker_rank
+        self.worker_handle_list = worker_handle_list
+        self.actor = ProxyActor.options(scheduling_strategy=scheduling_strategy).remote()
+
+        self.dtype = dtype
+        self.is_driver_worker = is_driver_worker
+        self.gpu_cache = gpu_cache
+
+        self.cache_device = "cpu"
+        self.num_migration_cache_blocks = self.migrate_config.migration_cache_blocks
+        self.num_layers = self.cache_engine.num_layers
+        self.migration_cache_size = self.cache_engine.block_size * self.cache_engine.num_heads * self.cache_engine.head_size
+
+        self.dummy_cache = torch.empty(
+            size=(2*self.num_migration_cache_blocks, self.num_layers, self.migration_cache_size),
+            dtype=self.cache_engine.dtype,
+            device=self.cache_device,
+            pin_memory=True
+        )
+        self.migration_stream = torch.cuda.Stream()
+
+    def init_col(self, name, world_size, rank) -> None:
+        logger.info("create rpc migrate backend success.")
+
+    def destory_col(self) -> None:
+        logger.info("destory rpc migrate backend success.")
+
+    def warmup(self) -> None:
+        self_handle = self.worker_handle_list[self.worker_rank]
+        self.actor.exec_method.remote(self.is_driver_worker, self_handle, "do_send", None, [0])
+
+    def migrate_cache(self, src_handle, src_blocks: List[int], dst_blocks: List[int]) -> None:
```

Comment:
We can add some brief annotations for all these migration functions, like d2h, h2d, communication, otherwise it's a little hard to understand only through the function name.

Reply:
Add try, except here for failover, referencing the original codes, so is else where. We need to carefully handle exception wherever we call ray.get.